src/transaction/locator_sr.c

```diff
@@ -2541,6 +2542,19 @@ xlocator_fetch (THREAD_ENTRY * thread_p, OID * oid, int chn, LOCK lock,
 		   * rules. This object must be also locked. */
 		  object_locked = true;
 		}
+
+#if defined (SERVER_MODE)
+	      if (lock == SCH_M_LOCK)
+		{
+		  assert (OID_IS_ROOTOID (class_oid));
+		  assert (!OID_ISNULL (p_oid));
+
+		  if (is_active_transaction_server ())
+		    {
+		      log_append_schema_modification_lock (thread_p, p_oid);
```

Comment:
Between the moment that the `SCH_M_LOCK` on the `p_oid` is obtained and the moment that a first log record for the transaction is actually added to the log, there could be:
  - a large pause in time - let's say, in a worst case scenario
  - failed verification and conditions which would cause the actual modifying log records to not be added (eg: the DDL modifying transaction to fail to validate for some reason) - resulting in a false negative on the PTS side

These are both unlikely scenarios but they can occur.
The first bullet above has the potential to stall a lot of PTS client transactions that could - otherwise - to their job just fine. And as far as I see, there would be no race condition as to the consistency of the log because, once the object is locked on ATS, no other ATS transaction can change it
Isn't it better to just collect the information as to what objects have been locked in the transaction descriptor - similar to `tx_transient_class_registry m_modified_classes` - but more with the semantics of `tx_transient_class_registry m_locked_classes` - and only add the schema-modification-lock log record upon actually adding the first log record of that transaction - eg: around `prior_lsa_next_record_internal`?

Reply:
@wrlawodms @joohok 
I think some of the aspects discussed by you above regarding to [possible] multiple logging of schema modification log records might be avoided with a lazy approach as described above. Can also be left out as a possible improvement.

Reply:
It seems like this could be an optimization point. I think I understand the concerns and the proposed solution.

If waiting for class modifications takes time after a lock for a class is obtained through `locator_fetch()`, PTS might have no choice but to wait. 
However, I believe that logging actions should occur where they take place. 
I think the approach of appending schema modification lock log within the log appending function seems unconventional.

There might also be a way to manage this from the PTS replicator's perspective. Instead of locking immediately when meeting a schema modification lock log, it could wait until one or more logs for that transaction have been replicated.

It appears that there is room for optimization regarding the gap between locking and actual changes within the PTS. I'd like to pursue this during the optimization phase. What are your thoughts on how to proceed?
Thanks.

Reply:
> There might also be a way to manage this from the PTS replicator's perspective. Instead of locking immediately when meeting a schema modification lock log, it could wait until one or more logs for that transaction have been replicated.

True.
The implementation of lazy-application of log records on the PTS side would be to the same end effect - not stalling PTS client read-only transactions - and would even be more localized - only in the bookkeeping & acquiring locks implementation.